Zend/zend_operators.c

```diff
@@ -2503,11 +2516,22 @@ static void ZEND_FASTCALL increment_string(zval *str) /* {{{ */
 	int ch;
 
 	if (Z_STRLEN_P(str) == 0) {
+		zend_error(E_DEPRECATED, "Increment on non-alphanumeric string is deprecated");
+		if (EG(exception)) {
+			return;
+		}
 		zval_ptr_dtor_str(str);
 		ZVAL_CHAR(str, '1');
 		return;
```

Comment:
This is... cursed but good catch :D

Reply:
So I pushed something which fixes the immediate crashed, but I don't know *what* the behaviour should be.

Should it *still* increment the original variable? Or should it attempt to increment the converted value?

Reply:
In my opinion, the least surprising would be if it behaved as if the value was not overridden during error handling.

This would also match the behavior of `+=` in the same situation (see my previous comment).

In the `Z_STRLEN_P(str) == 0` case, it already behaves like this.

In the `!zend_string_only_has_ascii_alphanumeric(Z_STR_P(str))` case, we can do this:
``` diff
index e558a1d7d8..ed9bba26c5 100644
--- a/Zend/zend_operators.c
+++ b/Zend/zend_operators.c
@@ -2527,14 +2528,20 @@ static bool ZEND_FASTCALL increment_string(zval *str) /* {{{ */
        }
 
        if (UNEXPECTED(!zend_string_only_has_ascii_alphanumeric(Z_STR_P(str)))) {
+               zend_string *zstr = Z_STR_P(str);
+               GC_TRY_ADDREF(zstr);
+
                zend_error(E_DEPRECATED, "Increment on non-alphanumeric string is deprecated");
                if (EG(exception)) {
+                       GC_TRY_DELREF(zstr);
+                       if (!GC_REFCOUNT(zstr)) {
+                               efree(zstr);
+                       }
                        return false;
                }
-               /* A userland error handler can change the type from string to something else */
-               if (Z_TYPE_P(str) != IS_STRING) {
-                       return false;
-               }
+
+               zval_ptr_dtor(str);
+               ZVAL_STR(str, zstr);
        }
 
        if (!Z_REFCOUNTED_P(str)) {
```

Reply:
@arnaud-lb Nice catch! As long as we make sure there is no separation in the common case (i.e. not installing a crazy error handler :wink:) I agree that your solution is preferable.